### PR TITLE
Bug 1796189: baremetal: Add the image URL to the baremetal CR

### PR DESCRIPTION
**Pull request by imain**
This fixes the issue we're having with the download container not knowing
where to get the upstream image from.

### CODE REVIEW

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1460/

File: pkg/asset/manifests/openshift.go [hunk 1]
```diff
@@ -44,6 +46,12 @@ type Openshift struct {
 	FileList []*asset.File
 }
 
+// Used for templating the baremetal CR.
+type BaremetalTemplateData struct {
+	Baremetal     *baremetal.Platform
+	RhcosImageURL string
+}
+
```
Comment (hardys) [inline pkg/asset/manifests/openshift.go]: This should perhaps go into pkg/types/baremetal ?

**Comment by hardys** [comment]
Based on previous review feedback on other PRs I think we'll be asked to move the baremetal specific struct definition, but otherwise lgtm

File: pkg/asset/manifests/openshift.go (hunk 1 above)
Comment (stbenjam) [reply pkg/asset/manifests/openshift.go]: At the very least if it remains here it shouldn't be exported (e.g. just baremetalTemplateData), it's only ever used here.

**Comment by sadasu** [comment]
Can we take care of cache_url also within the same PR?

**Comment by stbenjam** [comment]
> Can we take care of cache_url also within the same PR?

What's the plan for cache URL? I thought we were going to drop it now since we have all the overrides directly in the installer (bootstrapOsImage, clusterOsImage)

**Comment by sadasu** [comment]
> > Can we take care of cache_url also within the same PR?
> 
> What's the plan for cache URL? I thought we were going to drop it now since we have all the overrides directly in the installer (bootstrapOsImage, clusterOsImage)

Yes. Wasn't aware how support for bootstrapOsImage, clusterOsImage made the need for cache_url unnecessary. So, ignore my question.

File: pkg/asset/manifests/openshift.go (hunk 1 above)
Comment (imain) [reply pkg/asset/manifests/openshift.go]: Right good point.  Thanks guys.

Comment (imain) [reply pkg/asset/manifests/openshift.go]: The other alternative would be to define this in pkg/types/baremetal and try to use it both here and in the bootstrap templating.

**Comment by imain** [comment]
> Based on previous review feedback on other PRs I think we'll be asked to move the baremetal specific struct definition, but otherwise lgtm

I made it so it's unexported.  I personally think that's the appropriate way since it's only used in the one place but we'll see.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1468/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1469/

**Comment by stbenjam** [comment]
/retitle Bug 1796189: baremetal: Add the image URL to the baremetal CR

**Comment by stbenjam** [comment]
@imain Looks like this needs a rebase

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1474/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1477/

File: pkg/asset/manifests/openshift.go [hunk 2]
```diff
@@ -45,6 +47,12 @@ type Openshift struct {
 	FileList []*asset.File
 }
 
+// Used for templating the baremetal CR.
+type baremetalTemplateData struct {
+	Baremetal     *baremetal.Platform
+	ProvisioningOSDownloadURL string
```
Comment (stbenjam) [inline pkg/asset/manifests/openshift.go]: These should be aligned to make go fmt happy

File: pkg/asset/manifests/openshift.go [hunk 3]
```diff
@@ -201,7 +209,11 @@ func (o *Openshift) Generate(dependencies asset.Parents) error {
 		assetData["99_cloud-creds-secret.yaml"] = applyTemplateData(cloudCredsSecret.Files()[0].Data, templateData)
 		assetData["99_role-cloud-creds-secret-reader.yaml"] = applyTemplateData(roleCloudCredsSecretReader.Files()[0].Data, templateData)
 	case baremetaltypes.Name:
-		assetData["99_baremetal-provisioning-config.yaml"] = applyTemplateData(baremetalConfig.Files()[0].Data, *installConfig.Config.Platform.BareMetal)
+		bmTemplateData := baremetalTemplateData{
+			Baremetal:     installConfig.Config.Platform.BareMetal,
+			ProvisioningOSDownloadURL: string(*new(rhcos.Image)),
```
Comment (stbenjam) [inline pkg/asset/manifests/openshift.go]: This will just be an empty string, you can see how workers does it:


https://github.com/openshift/installer/blob/c47dffde52e56a111ab7ce3d7c429d597cb15a74/pkg/asset/machines/worker.go#L161

https://github.com/openshift/installer/blob/c47dffde52e56a111ab7ce3d7c429d597cb15a74/pkg/asset/machines/worker.go#L163

https://github.com/openshift/installer/blob/c47dffde52e56a111ab7ce3d7c429d597cb15a74/pkg/asset/machines/worker.go#L214

Comment (stbenjam) [reply pkg/asset/manifests/openshift.go]: Also, once you add this as a dependency, the graph in the docs needs to be updated, there's a command at the bottom of https://github.com/openshift/installer/blob/master/docs/design/assetgeneration.md#dependency-graph

File: pkg/asset/manifests/openshift.go (hunk 2 above)
Comment (imain) [reply pkg/asset/manifests/openshift.go]: Doh!  I thought for sure I ran gofmt on it.  Thanks.

File: pkg/asset/manifests/openshift.go (hunk 3 above)
Comment (imain) [reply pkg/asset/manifests/openshift.go]: Ah OK I see.  Thanks.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1480/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1481/

**Comment by stbenjam** [comment]
Thanks, this looks good to me, and my local test (still running) does set the value in the CR correctly, so it should complete successfully now:

```yaml
    provisioningOSDownloadURL: http://172.22.0.1/images/rhcos-44.81.202001241431.0-openstack.x86_64.qcow2.gz?sha256=1697f25f2f1270ce80971200313a4831cd8cdd4228cdabafff404bd57690be2d
```

/assign @abhinavdahiya PTAL

File: docs/design/resource_dep.svg [hunk 4]
```diff
@@ -1,1949 +1,1728 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
  "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
-<!-- Generated by graphviz version 2.40.1 (0)
+<!-- Generated by graphviz version 2.30.1 (20180420.1509)
```
Comment (abhinavdahiya) [inline docs/design/resource_dep.svg]: move this out to a separate commit maybe?

Comment (abhinavdahiya) [reply docs/design/resource_dep.svg]: i.e the entire change to resource_dep.svg

File: pkg/asset/manifests/openshift.go [hunk 5]
```diff
@@ -45,6 +47,12 @@ type Openshift struct {
 	FileList []*asset.File
 }
 
+// Used for templating the baremetal CR.
+type baremetalTemplateData struct {
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/openshift.go]: why not move it closer to https://github.com/openshift/installer/blob/4bb7986830ec15de43f6f0b6a5f40d9b761925bf/pkg/asset/manifests/template.go#L77

or however we did the platform separation for the bootkube/bootstrap.ign templating.

**Comment by abhinavdahiya** [comment]
had few nits, otherwise mo major complaints.

**Comment by imain** [comment]
> had few nits, otherwise mo major complaints.

Thanks, think I cleaned them up.

**Comment by stbenjam** [comment]
/lgtm

Thanks all!!

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/1482/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1483/
